components/prism-csharp.js

```diff
@@ -1,79 +1,135 @@
+
+	var classNameInside = {
+		'keyword': keywords,
+		'punctuation': /[<>(),.:[\]]/
+	};
+
+	Prism.languages.csharp = Prism.languages.extend('clike', {
+		'keyword': keywords,
+		'string': [
+			{
+				pattern: /@("|')(?:\1\1|\\[\s\S]|(?!\1)[^\\])*\1/,
+				greedy: true
+			},
+			{
+				pattern: /("|')(?:\\.|(?!\1)[^\\\r\n])*?\1/,
+				greedy: true
+			}
+		],
+		'class-name': [
+			{
+				// (Foo bar, Bar baz, Foo[,,] bay, Foo<Bar, FooBar<Bar>> bax)
+				pattern: /\b[A-Z]\w*(?:\.\w+)*(?:<(?:[^>\n=;{}]+|>(?=\s*[,.>?[)]))+>)?(?:\.\w+)*(?:\[\s*(?:,\s*)*\])*(?=\s+\w+(?:\s*[=,;:{)\]]|\s+in))/,
+				inside: classNameInside
+			},
+			{
+				// [Foo]
+				pattern: /((?:^|\W)\[)[A-Z]\w*(?:\.\w+)*\b/,
+				lookbehind: true,
+				inside: {
+					'punctuation': /\./
+				}
+			},
+			{
+				// class Foo<A, B>
+				pattern: /(\b(?:class|enum|interface|struct)\s+)[A-Z]\w*(?:\s*<[^>]+>)?/,
+				lookbehind: true,
+				inside: {
+					'punctuation': /[<>,]/
+				}
+			},
+			{
+				// catch(Foo)
+				pattern: /(\bcatch\s*\()[A-Z]\w*(?:\.\w+)*\b/,
+				lookbehind: true,
+				inside: {
+					'punctuation': /\./
+				}
+			},
+			{
+				// where Foo
+				pattern: /(\bwhere\s+)[A-Z]\w*\b/,
+				lookbehind: true
+			},
+			{
+				// as Foo<A>, is Bar<B>
+				pattern: /(\b(?:is|as)\s+)[A-Z]\w*(?:\.\w+)*(?:<(?:[^>\n=;{}]+|>(?=\s*[,.>?[)]))+>)?(?:\.\w+)*(?:\[\s*(?:,\s*)*\])*/,
+				lookbehind: true,
+				inside: classNameInside
 			}
+		],
+		'number': /(?:\b0(?:x[\da-f_]*[\da-f]|b[01_]*[01])|\B\.\d+|\b\d+(?:_+\d+)*\.?\d*)(?:ul|[flu])?\b/i
+	});
+
+	Prism.languages.insertBefore('csharp', 'class-name', {
+		'type-expression': {
+			// default(Foo), typeof(Foo<Bar>)
+			pattern: /(\b(?:default|typeof)\(\s*)[A-Z]\w*(?:\.\w+)*(?:<(?:[^>\n=;{}]+|>(?=\s*[,.>?[]))+>)?(?:\.\w+)*(?:\[\s*(?:,\s*)*\])*(?=\s*\))/,
+			lookbehind: true,
+			inside: classNameInside,
+			alias: 'class-name'
 		},
-		{
-			// [Foo]
-			pattern: /(\[)[A-Z]\w*(?:\.\w+)*\b/,
+		'constructor-invocation': {
+			// new List<Foo<Bar[]>> { }
+			pattern: /(\bnew\s+)[A-Z]\w*(?:\.\w+)*(?:<(?:[^>\n=;{}]+|>(?=\s*[,.>?[)]))+>)?(?:\.\w+)*(?:\[\s*(?:,\s*)*\])*(?=\s*[[({])/,
 			lookbehind: true,
+			inside: classNameInside,
+			alias: 'class-name'
+		},
+		'return-type': {
+			// Foo<Bar> ForBar(), Foo IFoo.Bar() => 0
+			pattern: /\b[A-Z]\w*(?:\.\w+)*(?:<[^\n=;{]+?>(?:\.\w+)*)?(?:\[\s*(,\s*)*\])?(?=\s+(?:[A-Z]\w*\.)?(?:\w+(?:\s*<[^\n=;{]+?>)?\s*(?:[({]|=>)|this\s*\[))/,
```

Comment:
I feel like this is redundant with the first class-name pattern defined previously. Are those look-ahead actually needed? Are there cases where a "class name-like" word (for example a capitalized word `Foo`) can appear in C# while not being an actual class name?

Reply:
Properties, some fields, compile time constants (basically fields) and methods are written in Pascal-case. So yes, there are many places in which `Foo` could not be a class. E.g. `Foo.Bar`: `Bar` could be a property, field, method or subclass. There is no way of knowing from the syntax alone.

The `return-type` is a stand-alone pattern because it has to be executed before `generic-method` (which is the look ahead). I know that it might seem complicated but generic methods (`Foo<A, B<C, D, (E, F>, (G, H)> I<J, K, L>();`), explicite implementation (`Foo<A> IBar<B, C<D>>.Bar<E, F>();` and stuff like properties and indexers make things a little more complicated...

Reply:
Alright. As long as we have good reasons to do complicated stuff, I'm ok with that ^^